### PR TITLE
Add landcover

### PR DESCRIPTION
**Pull request by ClaireDons**
Added landcover to the carbon tracker workflow.

### CODE REVIEW

**Comment by BSchilperoort** [comment]
Your first PR! Awesome :partying_face: 

It's still a draft PR, if it's ready for review (which I assume it is because you asked me to review), you can click the "Ready for review" button to make it a normal pull request.

File: src/excited_workflow/source_datasets/land_cover.py [hunk 1]
```diff
@@ -1,19 +1,117 @@
...
+                if not (
+                    dims_check(dat, target_grid, "latitude")
+                    and dims_check(dat, target_grid, "longitude")
+                ):
+                    dat.close()
+                    ds = self.regrid(file, target_grid, variables)
+                    ds.to_netcdf(name)
+            else:
+                ds = self.regrid(file, target_grid, variables)
+                ds.to_netcdf(name)
```
Comment (BSchilperoort) [inline src/excited_workflow/source_datasets/land_cover.py:113]: I think you could refactor this to:

```suggestion
            if not name.is_file() or not valid_regrid_file(name, target_grid): 
                self.regrid(file, target_grid, variables).to_netcdf(name)
```
Where the validity check (dims, st_size) can be wrapped in a function such as `valid_regrid_file`

File: src/excited_workflow/source_datasets/land_cover.py [hunk 2]
```diff
@@ -1,19 +1,117 @@
...
+    Args:
+        data: DataArray containing the time values, e.g. ds["time"].
+
+    Returns:
+        Numpy array with a datetime64 dtype.
+    """
+    return np.array([np.datetime64(el) for el in data.to_numpy()])
+
+
+def dims_check(dat: xr.Dataset, target: xr.Dataset, var: str) -> bool:
```
Comment (BSchilperoort) [inline src/excited_workflow/source_datasets/land_cover.py:25]:
```suggestion
def check_dims(dat: xr.Dataset, target: xr.Dataset, var: str) -> bool:
```
I can be picky with names :wink:

File: src/excited_workflow/source_datasets/land_cover.py [hunk 3]
```diff
@@ -32,19 +130,21 @@ def load(
         """
         self.validate_variables(variables)
 
-        files = list(self.get_path().glob("*.nc"))
+        if target_grid is None:
+            msg = "target_grid is not optional for loading landcover data."
+            raise ValueError(msg)
+
+        path = self.get_path() / "preprocessed"
```
Comment (BSchilperoort) [inline src/excited_workflow/source_datasets/land_cover.py:137]:
```suggestion
        preprocessed_dir = self.get_path() / "preprocessed"
```
This is a more clear name I think.

File: src/excited_workflow/source_datasets/land_cover.py [hunk 4]
```diff
@@ -32,19 +130,21 @@ def load(
...
+        path.mkdir(parents=True, exist_ok=True)
+        self.preprocess(path, target_grid, variables)
+
+        files = list(path.glob("*.nc"))
         if len(files) == 0:
-            msg = f"No netCDF files found at path '{self.get_path()}'"
+            msg = f"No netCDF files found at path '{path}'"
```
Comment (BSchilperoort) [inline src/excited_workflow/source_datasets/land_cover.py:143]:
```suggestion
            msg = f"No netCDF files found in directory '{preprocessed_dir}'"
```

**Review by BSchilperoort — CHANGES_REQUESTED** [review]
I didn't run the code yet, but here are already my comments on the code. Just some small refactoring suggestions, nice work!

File: src/excited_workflow/source_datasets/land_cover.py (hunk 1 above)
Comment (BSchilperoort) [reply src/excited_workflow/source_datasets/land_cover.py:113]: Additionally, I could be nice to print something here so the user gets feedback during the long pre-processing.
E.g. 
```py
print(f"File '{name}' was not found or not valid for this target dataset. Regridding raw data to `target_grid`")
```

Comment (ClaireDons) [reply src/excited_workflow/source_datasets/land_cover.py]: Ok created a new function and adjusted this, ruff was unhappy with the long text though so I shortened it. If you think longer is clearer, I can change it back.

File: src/excited_workflow/source_datasets/land_cover.py [hunk 5]
```diff
@@ -22,7 +22,41 @@ def _cftime_to_datetime(data: xr.DataArray) -> np.ndarray:
     return np.array([np.datetime64(el) for el in data.to_numpy()])
 
 
-def dims_check(dat: xr.Dataset, target: xr.Dataset, var: str) -> bool:
+def regrid(
+    file: Path,
+    target_grid: xr.Dataset,
+    variables: list[str] | None = None,
+) -> xr.Dataset:
+    """Regrids one file to target dataset.
```
Comment (BSchilperoort) [inline src/excited_workflow/source_datasets/land_cover.py:30]:
```suggestion
    """Regrid a land cover netCDF file to a target dataset, and return the regridded dataset.```

File: src/excited_workflow/source_datasets/land_cover.py [hunk 6]
```diff
@@ -22,7 +22,41 @@ def _cftime_to_datetime(data: xr.DataArray) -> np.ndarray:
...
+
+    Args:
+        file: file to regrid
+        variables: List of variable names which should be downloaded.
+        target_grid: Grid to which the data should be regridded to.
```
Comment (BSchilperoort) [inline src/excited_workflow/source_datasets/land_cover.py:35]:
```suggestion
        target_grid: Dataset to which the data should be regridded to.
        variables: List of variable names which should be regridded.
```
they're the other way around in the function definition.

File: src/excited_workflow/source_datasets/land_cover.py [hunk 7]
```diff
@@ -22,7 +22,41 @@ def _cftime_to_datetime(data: xr.DataArray) -> np.ndarray:
...
+        Regridded Dataset.
+    """
+    ds = xr.open_dataset(file, chunks={"lat": 2000, "lon": 2000})
+
+    # Set time to middle of bounds.
+    time_coords = _cftime_to_datetime(ds["time_bounds"].mean(dim="bounds"))
+    ds = ds.drop("time_bounds")
+    ds["time"] = time_coords
+
+    ds = ds[["lccs_class"]]  # Only take the class variable.
```
Comment (BSchilperoort) [inline src/excited_workflow/source_datasets/land_cover.py:47]: The variables argument is not used here then, right? The variables are already limited to only lccs_class

File: src/excited_workflow/source_datasets/land_cover.py (hunk 2 above)
Comment (BSchilperoort) [inline src/excited_workflow/source_datasets/land_cover.py]: Looking at this again, the coordinates (and their values) are checked, right?
It's probably better to reflect that in the function definition and docstring
```suggestion
def coord_matches(dat: xr.Dataset, target: xr.Dataset, coord: str) -> bool:
```

File: src/excited_workflow/source_datasets/land_cover.py [hunk 8]
```diff
@@ -1,19 +1,128 @@
...
+
+    if variables is not None:
+        ds = ds[variables]
+
+    ds = ds.regrid.most_common(target_grid, time_dim="time")
+
+    return ds
+
+
+def check_dims(dat: xr.Dataset, target: xr.Dataset, var: str) -> bool:
```
Comment (BSchilperoort) [inline src/excited_workflow/source_datasets/land_cover.py:59]:
```suggestion
def coord_matches(dat: xr.Dataset, target: xr.Dataset, coord: str) -> bool:
```
On a second look, it's the coordinate (and the values) which are checked. It's probably better to reflect that in the definition & docstring

File: src/excited_workflow/source_datasets/land_cover.py [hunk 9]
```diff
@@ -1,19 +1,128 @@
...
+        """
+        self.validate_variables(variables)
+
+        files = list(self.get_path().glob("*.nc"))
+        if len(files) == 0:
+            msg = f"No netCDF files found at path '{self.get_path()}'"
+            raise FileNotFoundError(msg)
+
+        for file in files:
+            name = process_path / (file.stem + ".nc")
```
Comment (BSchilperoort) [inline src/excited_workflow/source_datasets/land_cover.py:118]:
```suggestion
            name = process_path / file.name
```
Should be OK right? The files are already .nc files (according to the glob above).

File: src/excited_workflow/source_datasets/land_cover.py [hunk 10]
```diff
@@ -1,19 +1,128 @@
...
+
+            if not name.is_file() or not valid_regrid_file(name, target_grid):
+                regrid(file, target_grid, variables).to_netcdf(name)
+                print(
+                    f"'{name}' not found or not valid for target dataset. Regridding."
```
Comment (BSchilperoort) [inline src/excited_workflow/source_datasets/land_cover.py:123]: Just FYI it's also possible to split strings over multiple lines like this:
```suggestion
                    f"'{name}' was not found, or does not match the target dataset. "
                    "Regridding the file to the target dataset."
```
You do have to then watch that the message doesn't get too long for the user's terminal width.

But the message is probably OK like it is.

**Review by BSchilperoort — APPROVED** [review]
Just some small comments :)

Feel free to merge the PR when you've resolved them :+1: 
(or we can do that together tomorrow)

File: src/excited_workflow/source_datasets/land_cover.py (hunk 7 above)
Comment (ClaireDons) [reply src/excited_workflow/source_datasets/land_cover.py:47]: I think I just took this from the demo but you're right doesn't apply here.

File: src/excited_workflow/source_datasets/land_cover.py (hunk 2 above)
Comment (ClaireDons) [reply src/excited_workflow/source_datasets/land_cover.py]: yep good point

File: src/excited_workflow/source_datasets/land_cover.py (hunk 9 above)
Comment (ClaireDons) [reply src/excited_workflow/source_datasets/land_cover.py:118]: .name is a lot better!

**Pull request merged**